lib/metrics.js

```diff
@@ -22,169 +14,56 @@ class MetricsEvent {
-      ])))
-      .pipe(new SerializeStream())
-      .pipe(this.logs.createWriteStream(this.allMetricsPath))
-    return deduplicateStream
-  }
-
-  async maybeCreateMetricsFile (moduleName) {
-    await maybeCreateFile(
-      this.#getMetricsFilePath(moduleName),
-      formatLog(
-        JSON.stringify({ totalJobsCompleted: 0, totalEarnings: '0' }) + '\n'
-      )
-    )
-  }
-}
-
-class DeduplicateStream extends Transform {
-  constructor () {
-    super({ objectMode: true })
-    this.last = null
-  }
-
-  _transform (obj, _, callback) {
-    const isChanged = this.last === null ||
-        Object
-          .entries(obj)
-          .some(([key, value]) => this.last[key] !== value)
-    if (isChanged) {
-      this.last = obj
-      callback(null, obj)
-    } else {
-      callback()
-    }
-  }
-}
-
-class TelemetryStream extends Transform {
-  constructor (moduleName) {
-    assert(moduleName, 'moduleName required')
-    super({ objectMode: true })
-    this.lastMetrics = null
-    this.moduleName = moduleName
-  }
-
-  _transform (metrics, _, callback) {
-    if (typeof this.lastMetrics?.totalJobsCompleted === 'number') {
+  // - Merges metrics from all modules
+  submit (moduleName, metrics) {
+    if (
+      typeof this.moduleMetrics.get(moduleName)?.totalJobsCompleted === 'number'
+    ) {
       writeClient.writePoint(
         new Point('jobs-completed')
-          .stringField('module', this.moduleName)
+          .stringField('module', moduleName)
           .intField(
             'value',
-            metrics.totalJobsCompleted - this.lastMetrics.totalJobsCompleted
+            metrics.totalJobsCompleted -
+              this.moduleMetrics.get(moduleName).totalJobsCompleted
           )
       )
     }
-    this.lastMetrics = metrics
-    callback(null, metrics)
-  }
-}
-
-let mergedMetrics
-
-class MergeMetricsStream extends Transform {
-  constructor (allMetrics, moduleMetrics) {
-    assert.strictEqual(
-      typeof allMetrics?.totalJobsCompleted,
-      'number',
-      'allMetrics.totalJobsCompleted required'
-    )
-    assert.strictEqual(
-      typeof moduleMetrics?.totalJobsCompleted,
-      'number',
-      'moduleMetrics.totalJobsCompleted required'
-    )
-    super({ objectMode: true })
-    if (!mergedMetrics) {
-      mergedMetrics = allMetrics
+    this.moduleMetrics.set(moduleName, metrics)
+    const mergedMetrics = Object
+      .values(this.moduleMetrics)
+      .reduce((acc, metrics) => {
+        acc.totalJobsCompleted += metrics.totalJobsCompleted
+        acc.totalEarnings += metrics.totalEarnings
+        return acc
+      }, {
+        totalJobsCompleted: 0,
+        totalEarnings: 0
+      })
+    const isChanged = this.mergedMetrics === null ||
+      Object
+        .entries(this.mergedMetrics)
+        .some(([key, value]) => mergedMetrics[key] !== value)
+    if (isChanged) {
+      this.mergedMetrics = mergedMetrics
+      super.emit.call(this, 'update', mergedMetrics)
```

Comment:
Question: This will emit metrics as soon as the first module emitted metrics. That means that while not all modules have emitted their first metrics event, we will show a partial job count. Should we delay emitting metric events until all modules have submitted? Risk: If one fails, we will never emit metrics. Also: I think this is something we can iterate on, and keep as is for now

Reply:
Let's iterate on this.

I suggest requiring all modules to emit the initial job count on startup. 
- We can handle this inside Core for modules exposing HTTP API, we just need to make the first `GET /stats` call immediately after the module starts accepting HTTP connections
- For modules running via Zinnia, we can easily improve `zinniad` to report the initial job count at startup - see https://github.com/filecoin-station/zinnia/blob/2b592f8ac636352ed3f1d6f43b771b5547496565/daemon/station_reporter.rs#L25-L40

Reply:
I don't think this is going to be good enough: If a module behaves badly, then it still won't report the metrics right away  (even if we pull, via the http interface).

Reply:
Sure. If we want to handle modules behaving badly, then we need a more robust solution.

I think my proposal is super easy to implement (and maintain) and will significantly improve the happy path. ATM, Zinnia does not update the job counter until the first job is finished, which takes around a second in my testing.

Anyhow, I think the current implementation is good enough for this pull request and we can iterate on improvements later.

Reply:
Opened a PR in Zinnia: https://github.com/filecoin-station/zinnia/pull/230